File: train_colab_model.py
import os
import json
import numpy as np
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.metrics import accuracy_score, mean_squared_error
from sklearn.model_selection import train_test_split

with open('training_samples.json', 'r', encoding='utf-8') as f:
    data = json.load(f)


def extract_features(sample):
    text = f"{sample['category']}: {sample['description']}"
    text_lower = text.lower()
    words = text_lower.split()
    high_words = ['dangerous', 'urgent', 'emergency', 'flooding', 'damage', 'broken']
    med_words = ['concern', 'problem', 'issue', 'repair']
    low_words = ['small', 'minor', 'cosmetic']
    high_count = sum(1 for w in high_words if w in text_lower)
    med_count = sum(1 for w in med_words if w in text_lower)
    low_count = sum(1 for w in low_words if w in text_lower)
    categories = ['Pothole', 'Street Light', 'Water Leak', 'Drainage', 'Sidewalk', 'Traffic Signal', 'Waste']
    category_text = text.split(':')[0].strip()
    category_index = categories.index(category_text) if category_text in categories else -1
    cat_features = [1 if i == category_index else 0 for i in range(len(categories))]
    return np.array([high_count / 10, med_count / 10, low_count / 10, len(words) / 100, len(text) / 500, *cat_features], dtype=float)

X = np.array([extract_features(sample) for sample in data])
y_priority = [sample['correctedPriority'] for sample in data]
y_severity = [sample['correctedSeverity'] for sample in data]
y_risk = [sample['correctedRiskFactor'] for sample in data]

X_train, X_test, y_p_train, y_p_test = train_test_split(X, y_priority, test_size=0.2, random_state=42, stratify=y_priority)
_, _, y_s_train, y_s_test = train_test_split(X, y_severity, test_size=0.2, random_state=42)
_, _, y_r_train, y_r_test = train_test_split(X, y_risk, test_size=0.2, random_state=42)

priority_model = LogisticRegression(multi_class='multinomial', solver='lbfgs', max_iter=1000)
severity_model = LinearRegression()
risk_model = LinearRegression()
priority_model.fit(X_train, y_p_train)
severity_model.fit(X_train, y_s_train)
risk_model.fit(X_train, y_r_train)

os.makedirs('colab_model_artifacts', exist_ok=True)
artifact = {
    'priority': {
        'classes': priority_model.classes_.tolist(),
        'coefficients': priority_model.coef_.tolist(),
        'intercepts': priority_model.intercept_.tolist(),
    },
    'severity': {
        'coefficients': severity_model.coef_.tolist(),
        'intercept': severity_model.intercept_.tolist()[0],
    },
    'risk': {
        'coefficients': risk_model.coef_.tolist(),
        'intercept': risk_model.intercept_.tolist()[0],
    },
}
with open('colab_model_artifacts/model_artifact.json', 'w', encoding='utf-8') as f:
    json.dump(artifact, f, indent=2)

print('Priority accuracy:', round(accuracy_score(y_p_test, priority_model.predict(X_test)), 3))
print('Severity RMSE:', round(mean_squared_error(y_s_test, severity_model.predict(X_test)), 3))
print('Risk RMSE:', round(mean_squared_error(y_r_test, risk_model.predict(X_test)), 3))
print('Artifact written to colab_model_artifacts/model_artifact.json')

File: train_colab_model.mjs
import fs from 'fs';
import path from 'path';

const data = JSON.parse(fs.readFileSync(path.resolve('training_samples.json'), 'utf8'));

function extractFeatures(sample) {
  const text = `${sample.category}: ${sample.description}`;
  const textLower = text.toLowerCase();
  const words = textLower.split(/\s+/);
  const highWords = ['dangerous', 'urgent', 'emergency', 'flooding', 'damage', 'broken'];
  const medWords = ['concern', 'problem', 'issue', 'repair'];
  const lowWords = ['small', 'minor', 'cosmetic'];

  const highCount = highWords.reduce((count, word) => count + (textLower.includes(word) ? 1 : 0), 0);
  const medCount = medWords.reduce((count, word) => count + (textLower.includes(word) ? 1 : 0), 0);
  const lowCount = lowWords.reduce((count, word) => count + (textLower.includes(word) ? 1 : 0), 0);

  const categories = ['Pothole', 'Street Light', 'Water Leak', 'Drainage', 'Sidewalk', 'Traffic Signal', 'Waste'];
  const categoryText = (sample.category || '').trim();
  const categoryIndex = categories.indexOf(categoryText);
  const catFeatures = categories.map((_, index) => (index === categoryIndex ? 1 : 0));

  return [
    highCount / 10,
    medCount / 10,
    lowCount / 10,
    words.length / 100,
    text.length / 500,
    ...catFeatures,
  ];
}

function dot(a, b) {
  return a.reduce((sum, value, index) => sum + value * b[index], 0);
}

function trainLinearRegression(features, targets) {
  const weights = Array(features[0].length).fill(0);
  let bias = 0;
  const learningRate = 0.01;
  const epochs = 3000;

  for (let epoch = 0; epoch < epochs; epoch += 1) {
    let gradW = Array(weights.length).fill(0);
    let gradB = 0;

    for (let i = 0; i < features.length; i += 1) {
      const prediction = bias + dot(weights, features[i]);
      const error = prediction - targets[i];
      gradB += error;
      for (let j = 0; j < weights.length; j += 1) {
        gradW[j] += error * features[i][j];
      }
    }

    const scale = 2 / features.length;
    for (let j = 0; j < weights.length; j += 1) {
      weights[j] -= learningRate * gradW[j] * scale;
    }
    bias -= learningRate * gradB * scale;
  }

  return { coefficients: weights, intercept: bias };
}

function trainPriorityModel(features, priorities) {
  const classes = ['High', 'Medium', 'Low'];
  const classFeatures = classes.map((label) => features.filter((_, index) => priorities[index] === label));
  const coefficients = classFeatures.map((group) => {
    const mean = Array(features[0].length).fill(0);
    if (group.length === 0) {
      return mean;
    }
    group.forEach((vector) => {
      vector.forEach((value, index) => {
        mean[index] += value;
      });
    });
    return mean.map((value) => value / group.length);
  });

  return {
    classes,
    coefficients,
    intercepts: Array(classes.length).fill(0),
  };
}

const features = data.map(extractFeatures);
const priorities = data.map((sample) => sample.correctedPriority);
const severities = data.map((sample) => sample.correctedSeverity);
const risks = data.map((sample) => sample.correctedRiskFactor);

const priorityModel = trainPriorityModel(features, priorities);
const severityModel = trainLinearRegression(features, severities);
const riskModel = trainLinearRegression(features, risks);

const artifact = {
  priority: priorityModel,
  severity: severityModel,
  risk: riskModel,
};

const outDir = path.resolve('colab_model_artifacts');
fs.mkdirSync(outDir, { recursive: true });
fs.writeFileSync(path.join(outDir, 'model_artifact.json'), JSON.stringify(artifact, null, 2));

console.log('Trained model artifact written to colab_model_artifacts/model_artifact.json');
